types/providers.d.ts

```diff
@@ -114,15 +114,15 @@ interface CredentialInput {
   placeholder?: string
 }
 
-interface CredentialsConfig<C extends Record<string, CredentialInput> = {}>
+interface CredentialsConfig<C extends Record<string, CredentialInput> = {}, K extends keyof C = keyof C>
   extends CommonProviderOptions {
   type: "credentials"
   credentials: C
-  authorize(credentials: Record<keyof C, string>, req: NextApiRequest): Awaitable<User | null>
+  authorize(credentials: Record<K, string>, req: NextApiRequest): Awaitable<User | null>
 }
 
-export type CredentialsProvider = (
-  options: Partial<CredentialsConfig>
+export type CredentialsProvider = <C extends Record<string, CredentialInput>>(
+  options: Partial<CredentialsConfig<C, keyof C>>
```

Comment:
```suggestion
export type CredentialsProvider = <C extends Record<string, CredentialInput>>(
  options: Partial<CredentialsConfig<C>>
) => CredentialsConfig<C>
```

Reply:
I'm really surprised. Instantiate here this way was first step I did, as its the main problem and it was not enough... But if you say it works, then it works... :+1: